tests/solutions/test_models.py

```diff
@@ -293,7 +295,32 @@ def test_duplicated_files(self):
 
         files = {'Hello.java': 'class Hello {}', 'Hello.java': 'class Hello {}'}
         json_data = {'md5': self.valid_md5_hash, 'files': files}
-        checkpoint = Checkpoint.objects.create_checkpoint(
+        checkpoint = Checkpoint.objects.sync_checkpoint(
             json_data=json_data, task=self.task, user=self.alice,
         )
         self.assertEqual(checkpoint.checkpointfile_set.count(), 1)
+
+    def test_only_store_last_checkpoint(self):
+        """Test, that the editor discards outdated checkpoints."""
+
+        files = {'Hello.java': 'class Hello {}', 'Bye.java': 'class Bye {}'}
+        json_data = {'md5': self.valid_md5_hash, 'files': files}
+        outdated_checkpoint = Checkpoint.objects.sync_checkpoint(
+            json_data=json_data, task=self.task, user=self.alice,
+        )
+        outdated_checkpoint_files = outdated_checkpoint.checkpointfile_set.all()
+
+        # Update Hello.java
+        files['Hello.java'] = 'class Hello {public static void main(String[] args) {}}'
+        json_data = {'md5': self.valid_md5_hash_updated, 'files': files}
+        Checkpoint.objects.sync_checkpoint(
+            json_data=json_data, task=self.task, user=self.alice,
+        )
+
+        # The outdated checkpoint and its associated
+        # checkpoint files should have been deleted
+        with self.assertRaises(ObjectDoesNotExist):
+            Checkpoint.objects.get(pk=outdated_checkpoint.pk)
+        for outdated_checkpoint_file in outdated_checkpoint_files:
+            with self.assertRaises(ObjectDoesNotExist):
+                CheckpointFile.objects.get(pk=outdated_checkpoint_file.pk)
```

Comment:
I would prefer to not use the exception approach and to use a filter query instead of a loop, e.g.:

```python
self.assertFalse(Checkpoint.objects.exists(id=old_id))
self.assertFalse(CheckpointFile.objects.filter(checkpoint_id=old_id))
```

The latter works because empty querysets, like empty lists, evaluate to `False`. This is easier to read and faster.